### PR TITLE
Better error handling when running scripts

### PR DESCRIPTION
**Pull request by richfitz**
See https://github.com/mrc-ide/orderly3/pull/19 for use in orderly

This PR does a number of things to improve error handling:

* expands the "source" like function to better cope with errors, warning and optionally also echoing to the console. No longer over-agressively swallows errors
* update outpack_packet_end to finish but **not** insert the package - this allows us to return/write the metadata and to flush the logs. Note that I am currently not super happy with where we write out the metadata and am open for suggestions here
* a bit of docs around subtleties of "running" a script
* laying the groundwork for eventually having support for doing this in a separate process - in particular the interface to `evaluate_script` (input args and return list) should be doable via callr

There's a bit of a churn in things like the device stack checking but the logic is the same, difference is just where we throw the error.

### CODE REVIEW

File: R/evaluate.R [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+      sink(NULL) # nolint
+    }
+    msg <- ngettext(check,
+                    "Script left 1 sink open",
+                    sprintf("Script left %d sinks open", check))
+  } else {
+    msg <- sprintf("Script closed %d more sinks than it opened!", abs(check))
+  }
+  list(success = success, message = msg)
+}
```
Comment (r-ash) [inline R/evaluate.R:142]: This looks like it was well fiddly to work out! :)

**Review by r-ash — APPROVED** [review]
This looks good

File: R/evaluate.R (hunk 1 above)
Comment (richfitz) [reply R/evaluate.R:142]: this bit was fine actually - and copied over from orderly really - in this PR it just moved from the support.R file and no longer throws

File: R/packet.R [hunk 2]
```diff
@@ -116,11 +128,58 @@ outpack_packet_end <- function(packet) {
                                   file_hash = packet$files$immutable,
                                   file_ignore = packet$files$ignored,
                                   hash_algorithm = hash_algorithm)
-  outpack_insert_packet(packet$path, json, packet$root)
+  if (insert) {
+    outpack_insert_packet(packet$path, json, packet$root)
+  } else {
+    ## TODO: I am not sure what the best filename to use here is -
```
Comment (r-ash) [inline R/packet.R:134]: I think `outpack.json` sounds fine to me, it's in a directory with name as the id already right? And there will only ever be 1 inside that directory?

Comment (richfitz) [reply R/packet.R:134]: in orderly it will be in a directory with the id, but not in outpack generally (the source directory can be anywhere for that)

**Pull request merged**